Add more info to input field tooltips

emlys opened the pull request:
## Description
Fixes #854

* Update the RST-generating code in `spec_utils.py` to include an anchor for each arg. This can then be linked to in the HTML.
* Rename the `userguide_html` property to `userguide` for consistency (I think we just didn't do this before because with the workbench in a separate repo, it would have been tedious)
* Pass the `userguide` property down to the `AboutModal` component, and use it to add an external link to the user's guide entry for each arg.
* Refactor the `ArgInput` `render` function to avoid duplicating the new `userguide` stuff
* Some linting


## Checklist
- [ ] ~Updated HISTORY.rst (if these changes are user-facing)~
- [ ] Updated the user's guide (if needed)
- [ ] Tested the affected models' UIs (if relevant)

emlys commented on workbench/src/renderer/components/SetupTab/ArgInput/index.jsx:
```diff
@@ -8,7 +8,10 @@ import Col from 'react-bootstrap/Col';
+import { MdFolderOpen, MdInfo, MdOpenInNew } from 'react-icons/md';
+import { shell } from 'electron';
+
+const baseUserguideURL = 'https://storage.googleapis.com/natcap-dev-build-artifacts/invest/emlys/3.9.1.post3250+g130227a76/userguide';
```
So that you can try it out, I've left this pointing to a dev build that has the new section anchors. I'll change this to the standard `https://storage.googleapis.com/releases.naturalcapitalproject.org/invest-userguide/latest` before merging.

davemfish replied:
@emlys Oh I almost forgot about this. How should we handle it? Do we need to update the UG to build the anchors before replacing this URL and merging?

emlys commented:
Hey @davemfish I'm gonna say this is ready for an initial review. 
Note about testing: there is a test that the user's guide link exists, but opening it is not tested. Clicking the link in the test throws an error. Let's save that for https://github.com/natcap/invest/issues/849.

davemfish reviewed (changes requested):
Hi @emlys , thanks for doing this one! I just have a few minor comments in the code. I think the refactor of `ArgInput` is a nice improvement.

Looking at a few models, I have some mixed feelings about displaying the units.

I think they are really important for number inputs like these from CV:
> Model Resolution (m)
Maximum Fetch Distance (m)

I think they're potentially confusing for raster inputs. I worry some might read this as a suggestion that the input value itself should be a number. 
> Land Use/Land Cover (integer)
Evapotranspiration (mm)

Then there's a few units that I feel like don't add a ton of value and could just be a point of confusion:
>Cell Size (linear_unit) <-- Rec
Start Year (year) <-- Rec
Z Parameter (unitless) <-- AWY
Threshold Flow Accumulation (pixel) <-- DelineateIt

I'm not really sure what to do exactly...it seems most valuable to have the unit front-and-center for `number` inputs, except that these last ones are all `numbers` and I feel like the unit label on its own isn't super helpful; they really benefit from the About text.

Some ideas we can toss around:
* We could move all the units into the About modal, so people always see them in the context of the other text.
* We could leave the prominent labels for numbers only, but then we might want to consider removing some of the less-descriptive units altogether.
* To solve the "raster problem" We could move the units label into the placeholder text and add a "unit" label, so it's clear that the value represents the unit, not the type that belongs in the input field. For example: 
> _raster (units: mm)
number (units: unitless)
number (units: year)
number (units: pixel)_ <-- this is still confusing I think, unless we can make "pixel count".

emlys commented on workbench/src/renderer/components/SetupTab/ArgInput/index.jsx:
```diff
@@ -34,16 +37,40 @@ function filterSpatialOverlapFeedback(message, filepath) {
+        {/* units are null for unitless number types
+        because the 'units' property is there, but has no value */}
+        <b>{ units === null ? 'unitless' : units }</b>
+        <span>)</span>
```
It didn't work to have the parentheses loose in the `React.Fragment`, but I can use another `React.Fragment` to contain them, rather than a span.

emlys commented on workbench/src/renderer/components/SetupTab/ArgInput/index.jsx:
```diff
@@ -367,30 +385,48 @@ class AboutModal extends React.PureComponent {
+          </Modal.Header>
+          <Modal.Body>
+            {`${arg.about} `}
+            <a
```
Sure!

emlys commented on workbench/tests/renderer/setuptab.test.js:
```diff
@@ -125,11 +148,35 @@ describe('Arguments form input types', () => {
+    await findByRole('link', { name: /user guide/ });
+  });
+
+  test('expect units to be displayed for numbers', async () => {
```
Good point, I split these up into individual tests.

emlys commented:
> Then there's a few units that I feel like don't add a ton of value and could just be a point of confusion:

    Cell Size (linear_unit) <-- Rec
    Start Year (year) <-- Rec
    Z Parameter (unitless) <-- AWY
    Threshold Flow Accumulation (pixel) <-- DelineateIt

To address these unusual ones:
* Z Parameter (unitless): I personally like that it says "unitless", I think it's good to emphasize that, just like we emphasize the units for other numbers. And without it, we might get questions about what the units are. But I'm open to discuss if you think it's confusing.
* Threshold Flow Accumulation (pixel): While "pixel" is technically correct as a unit, I agree that "pixel count" or "number of pixels" would make more sense. This could be changed by giving it a custom format in spec_utils (there is already another special case that's handled this way.)
* Cell Size (linear_unit): Agree that this is confusing. Some options:
    * Give it a custom format so it would display "any linear unit" or something like that
    * Make it unitless, then describe in the `about` text that it can be any linear unit. (Potentially confusing if `(unitless)` is then displayed in the UI)
    * Make a new unit, `u.other`, for which no unit label at all is displayed in the UI.
* Start Year (year): I'm not sure that `year` is even the correct unit here. Since we're not measuring a length of time, it's more like an identifier. It could instead have `"units": None` or `"units":  u.other`.

emlys commented:
> To solve the "raster problem" We could move the units label into the placeholder text and add a "unit" label, so it's clear that the value represents the unit, not the type that belongs in the input field. For example: raster (units: mm)

This would work. Do you think the current formatting for `ratio` or `percent` or `integer` rasters is also confusing? For example: 
<img width="499" alt="image" src="https://user-images.githubusercontent.com/43770515/160485312-bae5f082-113f-4a8a-b052-551c3f3cf238.png">

emlys commented:
@davemfish, I think I've addressed everything from our discussion so far.
* No units labels for rasters.
* Added a new unit, `year_AD`, for inputs that are a particular year e.g. '2020'. No units label is displayed for these.
* Change the units label for `pixel` to 'number of pixels'
* Replace `linear_unit` with a new custom unit, `other`. No units label is displayed for these.
* No units label is displayed for unitless inputs.

emlys commented on workbench/tests/renderer/setuptab.test.js:
```diff
@@ -34,14 +34,13 @@ const BASE_ARGS_SPEC = {
+  // make a copy of each nested layer so we don't edit the original
   const spec = { ...BASE_ARGS_SPEC };
+  spec.args = { ...BASE_ARGS_SPEC.args };
+  spec.args.arg = { ...BASE_ARGS_SPEC.args.arg };
```
Oh that's a nice one liner!

emlys commented on workbench/src/renderer/components/SetupTab/ArgInput/index.jsx:
```diff
@@ -42,20 +42,19 @@ function FormLabel(props) {
+        <> (</>
+        <b>{ units }</b>
+        <>)</>
+      </>
```
That does look better with the arg name heavier and the units normal! And then we avoid that linting issue.

I agree that bold is too strong, but I might go heavier than `500` for the font weight. To my eye, `600` gives better contrast against the plain text. How's this look to you?
<img width="512" alt="image" src="https://user-images.githubusercontent.com/43770515/160922396-75b0b497-2d7f-4dd3-baa0-15d474be4e65.png">

davemfish replied:
Looks good to me!

emlys commented:
@davemfish the ReadTheDocs build is failing everywhere for a different reason, I suspect a dependency update. `make userguide` works locally though. So I think the UG is fine.

davemfish reviewed (approved):
Nice!

davemfish commented:
> So that you can try it out, I've left this pointing to a dev build that has the new section anchors. I'll change this to the standard https://storage.googleapis.com/releases.naturalcapitalproject.org/invest-userguide/latest before merging.


@emlys Oh I almost forgot about this. How should we handle it? Do we need to update the UG to build the anchors before replacing this URL and merging?

emlys commented:
@davemfish the UG also needs this PR in order to build with the new anchors. I think we merge this one first, then the next UG build will have the anchors. Then the links shoudl work. (If you follow a link to an anchor that doesn't exist yet, there's no error, it just takes you to the top of the page.) I just changed the base URL back to the default.
